diff -Nurpw --exclude=CMakeLists.txt --exclude=libcharls.a --exclude='*.o' --exclude='Makefile*' libcharls.orig/header.cc libcharls/header.cc
--- libcharls.orig/header.cc	2009-09-04 01:12:42.000000000 +0200
+++ libcharls/header.cc	2009-09-04 10:21:20.000000000 +0200
@@ -9,11 +9,6 @@
 #include "encoderstrategy.h"
 #include <memory>
 
-
-// JFIF\0
-BYTE jfifID[] = {'J','F','I','F','\0'};
-
-
 bool IsDefault(const JlsCustomParameters* pcustom)
 {
 	if (pcustom->MAXVAL != 0)
@@ -379,6 +374,7 @@ void JLSInputStream::ReadComment()
 //
 void JLSInputStream::ReadJfif()
 {
+	const BYTE jfifID[] = {'J','F','I','F','\0'};
 	for(int i = 0; i < (int)sizeof(jfifID); i++)
 	{
 		if(jfifID[i] != ReadByte())
@@ -406,6 +402,7 @@ void JLSInputStream::ReadJfif()
 //
 JpegMarkerSegment* CreateJFIF(const JfifParamaters* jfif)
 {
+	const BYTE jfifID[] = {'J','F','I','F','\0'};
 	std::vector<BYTE> rgbyte;
 	for(int i = 0; i < (int)sizeof(jfifID); i++)
 	{
@@ -524,7 +521,6 @@ void JLSOutputStream::AddScan(const void
 	_icompLast += 1;
 	_segments.push_back(EncodeStartOfScan(pparams,pparams->ilv == ILV_NONE ? _icompLast : -1));
 
-	Size size = Size(pparams->width, pparams->height);
 	int ccomp = pparams->ilv == ILV_NONE ? 1 : pparams->components;
 		_segments.push_back(new JpegImageDataSegment(pbyteComp, *pparams, _icompLast, ccomp));
 }
